src/backend/distributed/planner/multi_planner.c

```diff
@@ -294,176 +298,183 @@ CreateDistributedPlan(PlannedStmt *localPlan, Query *originalQuery, Query *query
+	if (multiPlan->masterQuery)
+	{
+		finalPlan = CreateRealTimeFinalPlan(localPlan, multiPlan, customScan);
+	}
+	else
+	{
+		finalPlan = CreateRouterFinalPlan(localPlan, multiPlan, customScan);
 	}
 
-	result->planTree = (Plan *) fauxFunctionScan;
+	return finalPlan;
+}
 
-	return result;
+
+static PlannedStmt *
+CreateRealTimeFinalPlan(PlannedStmt *localPlan, MultiPlan *multiPlan,
+						CustomScan *customScan)
+{
+	PlannedStmt *finalPlan = NULL;
+
+	finalPlan = MasterNodeSelectPlan(multiPlan, customScan);
+	finalPlan->queryId = localPlan->queryId;
+	finalPlan->utilityStmt = localPlan->utilityStmt;
+
+	return finalPlan;
 }
 
 
-/*
- * GetMultiPlanString returns either NULL, if the plan is not a distributed
- * one, or the string representing the distributed plan.
- */
-static char *
-GetMultiPlanString(PlannedStmt *result)
+static PlannedStmt *
+CreateRouterFinalPlan(PlannedStmt *localPlan, MultiPlan *multiPlan,
+					  CustomScan *customScan)
 {
-	FunctionScan *fauxFunctionScan = NULL;
-	RangeTblFunction *fauxFunction = NULL;
-	FuncExpr *fauxFuncExpr = NULL;
-	Const *multiPlanData = NULL;
+	PlannedStmt *finalPlan = NULL;
+	RangeTblEntry *customScanRangeTableEntry = NULL;
+	ListCell *targetEntryCell = NULL;
+	List *targetList = NIL;
+	List *columnNameList = NIL;
+
+	int rangeTableCount = list_length(localPlan->rtable);
+	int customScanRangeTableIndex = rangeTableCount + 1;
 
-	if (!IsA(result->planTree, FunctionScan))
+	/* build a targetlist to "read" from the custom scan output */
+	foreach(targetEntryCell, localPlan->planTree->targetlist)
 	{
-		return NULL;
-	}
+		TargetEntry *targetEntry = lfirst(targetEntryCell);
+		TargetEntry *newTargetEntry = NULL;
+		Var *newVar = NULL;
+		Value *columnName = NULL;
 
-	fauxFunctionScan = (FunctionScan *) result->planTree;
+		Assert(IsA(targetEntry, TargetEntry));
 
-	if (list_length(fauxFunctionScan->functions) != 1)
-	{
-		return NULL;
-	}
+		/*
+		 * This is unlikely to be hit because we would not need resjunk stuff
+		 * at the toplevel of a router query - all things needing it have been
+		 * pushed down.
+		 */
+		if (targetEntry->resjunk)
+		{
+			continue;
+		}
 
-	fauxFunction = linitial(fauxFunctionScan->functions);
+		/* build target entry pointing to custom scan range table entry */
+		newVar = makeVarFromTargetEntry(customScanRangeTableIndex, targetEntry);
+		newTargetEntry = flatCopyTargetEntry(targetEntry);
+		newTargetEntry->expr = (Expr *) newVar;
+		targetList = lappend(targetList, newTargetEntry);
 
-	if (!IsA(fauxFunction->funcexpr, FuncExpr))
-	{
-		return NULL;
+		columnName = makeString(targetEntry->resname);
+		columnNameList = lappend(columnNameList, columnName);
 	}
 
-	fauxFuncExpr = (FuncExpr *) fauxFunction->funcexpr;
+	customScan->scan.plan.targetlist = targetList;
+	customScanRangeTableEntry = CustomScanRangeTableEntry(columnNameList);
 
-	if (fauxFuncExpr->funcid != CitusExtraDataContainerFuncId())
-	{
-		return NULL;
-	}
+	finalPlan = localPlan;
```

Comment:
Hm. It might be better to create a "fresh" PlannedStmt here too, like we're doing for real-time/tt.  Right now the PlannedStmt basically has a lot of bogus fields set (like ctes which have been pushed down and such).